### DISCUSSION TITLE
How to perform distributed-mesh computation?

### DISCUSSION THREAD

**Discussion by zhang-rite** [idaholab/moose#17200] [Q&A General] [answered]
Hi MOOSE experts

 I am trying to modify a 2D porous media hydromechanical problem to 3D.  In 3D, the model generates the mesh using CartesianMeshGenerator and has less than 150 M elements. A direct run (e.g., mpiexec -n 50) is always out of memory. 
I have tried to split the mesh into 50 chunks by running:
```
mpiexec -n 50 ./porous_flow-opt -i test.i --split-mesh 50
```
Then run by

```
mpiexec -n 50 ./porous_flow-opt -i test.i --use-split
```
However, the memory-out problem remains. The problem file is here ([github link ](https://github.com/zhang-rite/MOOSE_learning/blob/main/test3d.i)). 
I am working on a university cluster. Each node has 56 cpus and a shared memory about 170 G.

Perhaps something wrong with my MOOSE installation?
The run_test log can be seen here ([github link](https://github.com/zhang-rite/MOOSE_learning/blob/main/test_log.txt))

Could someone give some solutions? Thanks!

**Comment by loganharbour** [comment]
You don’t need that many processes for the initial split. When splitting the mesh, just a few procs is sufficient. Give something like 5 a try. 

Keep in mind that during the splitting, if the mesh is replicated it must be loaded on each proc. In general terms, this means you could be effectively loading (150 million x 50 procs) elements! Instead, with 5 processors the mesh is duplicated only 5 times. 

Are you running out of memory splitting the mesh or running with the split mesh?

**Reply by zhang-rite** [reply]
> Are you running out of memory splitting the mesh or running with the split mesh?
The latter. There is no problem when splitting the mesh. I will soon try usingless procs. Thank you.

**Reply by loganharbour** [reply]
When you run the latter problem, do you get to the point where the framework prints out its information (version, number of elements, variables, etc)? If so, paste that here.

**Reply by zhang-rite** [reply]
Initializing equation system                                                               [  1.82 s]
Framework Information:
MOOSE Version:           git commit dfbdf0e on 2020-12-02
LibMesh Version:         0d19e76fc8e7345a0634cba450f488b3a5626eb3
PETSc Version:           3.13.3
SLEPc Version:           3.13.3
Current Time:            Wed Mar  3 23:34:35 2021
Executable Timestamp:    Tue Dec  8 09:28:56 2020

Parallelism:
  Num Processors:          50
  Num Threads:             1

Mesh: 
  Parallel Type:           distributed (pre-split)
  Mesh Dimension:          3
  Spatial Dimension:       3
  Nodes:                   
    Total:                 1521194
    Local:                 33277
  Elems:                   
    Total:                 1480800
    Local:                 29621
  Num Subdomains:          2
  Num Partitions:          1
  Partitioner:             parmetis

Nonlinear System:
  Num DOFs:                6084776
  Num Local DOFs:          133108
  Variables:               { "pwater" "disp_x" "disp_y" "disp_z" } 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 

Auxiliary System:
  Num DOFs:                60712800
  Num Local DOFs:          1214461
  Variables:               { "stress_xx" "stress_zz" "stress_yy" "strain_yy" "strain_yy_0" "strain_yy_2" 
                             "strain_xx" "strain_zz" "perm_md" "perm" } "PorousFlow_effective_fluid_pressure_qp" 
                             
  Finite Element Types:    "MONOMIAL" "MONOMIAL" 
  Approximation Orders:    "FIRST" "CONSTANT" 

Execution Information:
  Executioner:             Steady
  Solver Mode:             NEWTON
  PETSc Preconditioner:    lu 
  MOOSE Preconditioner:    SMP


Outputting exodus .................Warning:  This MeshOutput subclass only supports meshes which have been serialized!
Warning:  This MeshOutput subclass only supports meshes which have been serialized!
...Warning, Exodus files cannot have titles longer than 80 characters.  Your title will be truncated.
.........                                            [ 30.31 s]
Computing initial residual ...                                                             [  4.46 s]
Updating displaced mesh                                                                    [  1.69 s]
 0 Nonlinear |R| = 5.964885e+08
Updating displaced mesh                                                                    [  1.71 s]

**Reply by zhang-rite** [reply]
Then after several mins the system stopped the run. I received the information with "Memory size limit exceeded" from the cluster.

**Reply by permcody** [reply]
Depending on how big your mesh is, you may also need to switch your output to Nemesis (`DistributedMesh` output analog), otherwise you will end up serializing the mesh to the root processor for output, which could still run you out of memory. 

```
[Outputs]
 ...
 nemesis = true
 exodus = false
[]
```

**Reply by zhang-rite** [reply]
Thank you for this suggestion. This partially solves the problem.

**Comment by WilkAndy** [comment, marked as answer]
I would hazard a guess that it is your use of LU+MUMPS.  Although that might work for smaller meshes, you might be forced to try ASM+LU+MUMPS, or HYPRE+BOOMERAMG, or ILU, or something else - i'm sure lots of other people can suggest other things to try.

If this is the case, then.... when you try different solver choices, i strongly suggest you do it on a smaller problem, so you get more rapid feedback on whether you have made a good choice (pay attention to the rate of convergence of the linear and the nonlinear residual, and feel free to ask for more help if you're not entirely sure).  i recommended MUMPS in the PorousFlow doco because it is a "generally good" choice (works for most models) but there are plenty of models which behave very satisfactorily with a "weaker", less memory-intensive option.

a

**Reply by zhang-rite** [reply]
Andy, many thanks. Yes, the suggestion helps.  This 3D problem is essentially the same as the previously posted 2D problem. After your suggestion, I have taken some time on testing the setting of the solver and preconditioner. 
Basically, I find 
```
 petsc_options_value = 'bcgs bjacobi'

```
or 
```
 petsc_options_value = 'hypre boomeramg'
```
can reduce the residual. 
Now the problem is the low solver efficiency.
